results/cmd/watcher/main_test.go

```diff
@@ -0,0 +1,155 @@
+		lis.Close()
+		s.Stop()
+		conn.Close()
+	})
+	return pb.NewResultsClient(conn)
+}
+
+// TestReconcile tests if TaskRun in the client can be properly updated when sent to reconciler
+func TestReconcile(t *testing.T) {
+	client := newResultsClient(t)
+	taskRun := &v1beta1.TaskRun{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        "Tekton",
+			Namespace:   "default",
+			Annotations: map[string]string{"demo": "demo"},
+			UID:         "12345",
+		},
+	}
+
+	asset, ctx := getFakeClients(t, []*v1beta1.TaskRun{taskRun}, client)
+	// Create a new TaskRun in the fake client, completed TaskRun should be updated with a results id field in annotations.
+	t.Run("test-create", func(t *testing.T) {
+		tr, err := reconcileTaskRun(ctx, asset, taskRun)
+		if err != nil {
+			t.Fatalf("Failed to get completed TaskRun %s: %v", taskRun.Name, err)
+		}
+		if _, ok := tr.Annotations[idName]; !ok {
+			t.Fatalf("Expected completed TaskRun %s should be updated with a results_id field in annotations", taskRun.Name)
+		}
+		if _, err := client.GetTaskRunResult(ctx, &pb.GetTaskRunRequest{ResultsId: tr.Annotations[idName]}); err != nil {
+			t.Fatalf("Expected completed TaskRun %s not created in api server", taskRun.Name)
+		}
+	})
+
+	// Update a TaskRun with results id, completed TaskRun should remain unchanged.
+	t.Run("test-unchange", func(t *testing.T) {
+		tr, err := asset.Clients.Pipeline.TektonV1beta1().TaskRuns(taskRun.Namespace).Get(taskRun.Name, metav1.GetOptions{})
+		if err != nil {
+			t.Fatalf("Failed to get completed TaskRun %s: %v", taskRun.Name, err)
+		}
+		newtr, err := reconcileTaskRun(ctx, asset, tr)
+		if err != nil {
+			t.Fatalf("Failed to get completed TaskRun %s: %v", taskRun.Name, err)
+		}
+		if diff := cmp.Diff(tr, newtr); diff != "" {
+			t.Fatalf("Expected completed TaskRun should remain unchanged when it has a results_id in annotations: %v", diff)
+		}
+	})
+
+	// Update a TaskRun with new value and results id, completed TaskRun should be updated.
+	t.Run("test-update", func(t *testing.T) {
+		tr, err := reconcileTaskRun(ctx, asset, taskRun)
+		if err != nil {
+			t.Fatalf("Failed to get completed TaskRun %s: %v", taskRun.Name, err)
+		}
+		tr.UID = "234435"
+		asset.Clients.Pipeline.TektonV1beta1().TaskRuns(taskRun.Namespace).Update(tr)
+		updatetr, err := reconcileTaskRun(ctx, asset, tr)
+		if err != nil {
+			t.Fatalf("Failed to get completed TaskRun %s: %v", taskRun.Name, err)
+		}
+		if diff := cmp.Diff(tr, updatetr); diff != "" {
+			t.Fatalf("Expected completed TaskRun should be updated in cluster: %v", diff)
+		}
+		res, err := client.GetTaskRunResult(ctx, &pb.GetTaskRunRequest{ResultsId: tr.Annotations[idName]})
+		if err != nil {
+			t.Fatalf("Expected completed TaskRun %s not created in api server", taskRun.Name)
+		}
+		pr, err := convert.ToProto(updatetr)
+		if err != nil {
+			t.Fatalf("failed to convert to proto: %v", err)
+		}
+		if diff := cmp.Diff(pr.String(), res.GetTaskRun().String()); diff != "" {
```

Comment:
We should compare the structs directly, not the strings. Theres a chance that the string function may not contain all the information needed to check for equality.
```suggestion
		if diff := cmp.Diff(pr, res.GetTaskRun()); diff != "" {
```

Reply:
got it, by the way, to compare unexported struct like proto, like in test for api, we have to use ``diff := cmp.Diff(pr, res.GetTaskRun(), protocmp.Transform())``

Reply:
done